AC_PrecLand: Add LANDING_TARGET_TYPE and LANDING_TARGET.type support

fnoop opened the pull request:
Add preliminary support for LANDING_TARGET.type field.  This type field represents the target type, defined in LANDING_TARGET_TYPE:
```
Field Name | Description
-- | --
0 | LANDING_TARGET_TYPE_LIGHT_BEACON | Landing target signaled by light beacon (ex: IR-LOCK)
1 | LANDING_TARGET_TYPE_RADIO_BEACON | Landing target signaled by radio beacon (ex: ILS, NDB)
2 | LANDING_TARGET_TYPE_VISION_FIDUCIAL | Landing target represented by a fiducial marker (ex: ARTag)
3 | LANDING_TARGET_TYPE_VISION_OTHER | Landing target represented by a pre-defined visual shape/feature (ex: X-marker, H-marker, square)
```
This PR doesn't do anything yet, other than add target_type to AC_PrecLand, AC_PrecLand_Backend, and to return 0 (Light Beacon) for IRLock and SITL_Gazebo backends, 1 (Radio Beacon) for SITL backend, and take the value from Mavlink LANDING_TARGET.type field for Companion backend.

It also adds a TgtType log field to precland PL log entries.

khancyr commented on libraries/AC_PrecLand/AC_PrecLand.h:
```diff
@@ -74,6 +74,9 @@ class AC_PrecLand
     bool target_acquired();
 
+    // returns target type
+    uint8_t target_type() const { return _target_type; }
```
LANDING_TARGET_TYPE

khancyr commented on libraries/AC_PrecLand/AC_PrecLand_Companion.cpp:
```diff
@@ -54,6 +55,12 @@ float AC_PrecLand_Companion::distance_to_target()
 }
 
+// return target type
+uint8_t AC_PrecLand_Companion::target_type()
```
LANDING_TARGET_TYPE

khancyr commented on libraries/AC_PrecLand/AC_PrecLand_Companion.h:
```diff
@@ -34,6 +34,9 @@ class AC_PrecLand_Companion : public AC_PrecLand_Backend
     float distance_to_target() override;
 
+    // returns target type
+    uint8_t target_type() override;
```
LANDING_TARGET_TYPE

khancyr commented on libraries/AC_PrecLand/AC_PrecLand_Companion.h:
```diff
@@ -44,4 +47,6 @@ class AC_PrecLand_Companion : public AC_PrecLand_Backend
     bool                _have_los_meas;         // true if there is a valid measurement from the camera
     uint32_t            _los_meas_time_ms;      // system time in milliseconds when los was measured
+    
+    uint8_t             _target_type;           // LANDING_TARGET_TYPE enum
```
LANDING_TARGET_TYPE

khancyr commented on libraries/AC_PrecLand/AC_PrecLand_IRLock.cpp:
```diff
@@ -54,3 +54,9 @@ uint32_t AC_PrecLand_IRLock::los_meas_time_ms() {
+uint8_t AC_PrecLand_IRLock::target_type()
+{
+    return 0;
+}
```
miss newline at end of file

rmackay9 commented:
.. but for the companion computer case, I guess the CC will be sending in these landing-target messages to ArduPilot so we could just store the values coming in from the CC to a local variable, no need to have a parameter.  TBH, I'm not really sure why the GCS would care about what type of sensor is used for the precision landing but getting it right in the message to the GCS can't be bad.  Adding another parameter for it though doesn't seem super useful to me though..

fnoop commented:
Refactored and squashed into two commits - one for AC_Precland and one for Copter.
@rmackay9 Sorry my bad - I didn't realise AP_int8 signified a parameter - this PR doesn't add or involve another parameter in any way.

uavosky commented:
> .. but for the companion computer case, I guess the CC will be sending in these landing-target messages to ArduPilot so we could just store the values coming in from the CC to a local variable, no need to have a parameter. TBH, I'm not really sure why the GCS would care about what type of sensor is used for the precision landing but getting it right in the message to the GCS can't be bad. Adding another parameter for it though doesn't seem super useful to me though..

Yep, but in my case it helpful ho have such us parameter. is start point to implement code for different beacon's. For example, radio-based beacon receivers (like standart ILS system with emmitors on the ground) have information about NE offset (relative to the vehicle) and information not depend of body state, so estimator shuld not correct it according 'camera' and body orientation. Or it can give to AP ability to receive precalculated position shift form companion computer. All i said IMHO (i'm new in AP code so i can miss some implemented functionality :) )

OXINARF commented on libraries/AC_PrecLand/AC_PrecLand.h:
```diff
@@ -128,6 +131,7 @@ class AC_PrecLand
     AP_Float                    _accel_noise;       // accelometer process noise
     AP_Vector3f                 _cam_offset;        // Position of the camera relative to the CG
 
+    LANDING_TARGET_TYPE         _target_type;       // target type (LANDING_TARGET_TYPE enum)
```
Is there a reason why this needs to be saved instead of the method directly asking the backend when the frontend method is called?

fnoop replied:
Connected frontend target_type() to backend target_type()

OXINARF commented on libraries/AC_PrecLand/AC_PrecLand_IRLock.h:
```diff
@@ -36,11 +36,15 @@ class AC_PrecLand_IRLock : public AC_PrecLand_Backend
     Vector3f            _los_meas_body;         // unit vector in body frame pointing towards target
     bool                _have_los_meas;         // true if there is a valid measurement from the camera
     uint32_t            _los_meas_time_ms;      // system time in milliseconds when los was measured
+    LANDING_TARGET_TYPE _target_type;           // LANDING_TARGET_TYPE enum
```
Unused.

fnoop replied:
Connected frontend target_type() to backend target_type()

OXINARF commented on libraries/AC_PrecLand/AC_PrecLand_SITL.h:
```diff
@@ -32,10 +32,14 @@ class AC_PrecLand_SITL : public AC_PrecLand_Backend
 
     Vector3f            _los_meas_body;         // unit vector in body frame pointing towards target
     uint32_t            _los_meas_time_ms;      // system time in milliseconds when los was measured
+    LANDING_TARGET_TYPE _target_type;           // LANDING_TARGET_TYPE enum
```
Unused.

fnoop replied:
Connected frontend target_type() to backend target_type()

OXINARF commented on libraries/AC_PrecLand/AC_PrecLand_SITL_Gazebo.h:
```diff
@@ -34,12 +34,16 @@ class AC_PrecLand_SITL_Gazebo : public AC_PrecLand_Backend
     Vector3f            _los_meas_body;         // unit vector in body frame pointing towards target
     bool                _have_los_meas;         // true if there is a valid measurement from the camera
     uint32_t            _los_meas_time_ms;      // system time in milliseconds when los was measured
+    LANDING_TARGET_TYPE _target_type;           // LANDING_TARGET_TYPE enum
```
Unused.

fnoop replied:
Connected frontend target_type() to backend target_type()

amilcarlucas commented:
This needs a rebase

khancyr commented:
Rework in #17564
